File: src/rust/src/roman_to_int_test.rs
// https://leetcode.com/problems/roman-to-integer
use std::collections::HashMap;

pub fn solution(s: String) -> i32 {
    let chars = s.chars().collect::<Vec<char>>();
    let mut set = HashMap::new();
    set.insert('I', 1);
    set.insert('V', 5);
    set.insert('L', 50);
    set.insert('X', 10);
    set.insert('C', 100);
    set.insert('D', 500);
    set.insert('M', 1000);

    let mut result = 0;
    for i in 0..s.len() {
        if i + 1 < chars.len() && set.get(&chars[i]) < set.get(&chars[i + 1]) {
            result -= set.get(&chars[i]).unwrap();
        } else {
            result += set.get(&chars[i]).unwrap();
        }
    }

    result
}

File: src/challenges/python/problems/roman_to_int_test.py
# https://leetcode.com/problems/roman-to-integer

from core.problem_base import *

class RomanToInt(ProblemBase):
    def Solution(self, s: str) -> int:
        map = { 
            'I': 1,
            'V': 5,
            'L': 50,
            'X': 10,
            'C': 100,
            'D': 500,
            'M': 1000,
        }

        result = 0
        for i in range(len(s)):
            if i + 1 < len(s) and map[s[i]] < map[s[i + 1]]:
                result -= map[s[i]] 
            else:
                result += map[s[i]]

        return result

if __name__ == '__main__':
    TestGen(RomanToInt) \
        .Add(lambda tc: tc.Param("III").Result(3)) \
        .Add(lambda tc: tc.Param("LVIII").Result(58)) \
        .Add(lambda tc: tc.Param("MCMXCIV").Result(1994)) \
        .Add(lambda tc: tc.Param("MMMCDXC").Result(3490)) \
        .Run()
